server/src/utils.ts

```diff
@@ -64,6 +65,131 @@ export async function validateImagePaths(textDocument: TextDocument, xmlData: Do
   return diagnostics
 }
 
+export async function validateLinks(xmlData: Document, knownModules: string[]): Promise<Diagnostic[]> {
+  const diagnostics: Diagnostic[] = []
+
+  diagnostics.push(...validateSamePageLinks(xmlData, knownModules))
+  diagnostics.push(...validateOtherPageLinks(xmlData, knownModules))
+
+  return diagnostics
+}
+
+export async function getCurrentModules(workspaceFolders: WorkspaceFolder[]): Promise<string[]> {
+  const moduleFiles: string[] = []
+
+  for (const workspace of workspaceFolders) {
+    const modulesPath = path.join(URI.parse(workspace.uri).path, 'modules')
+    if (fs.existsSync(modulesPath)) {
+      // Return modules with full path information in case we need to peek into
+      // any of them later (e.g. for validation purposes)
+      moduleFiles.push(
+        ...glob.sync('**/*.cnxml', { cwd: modulesPath }).map(
+          (val) => { return modulesPath.concat('/', val) }
+        )
+      )
+    }
+  }
+
+  return moduleFiles
+}
+
+function validateOtherPageLinks(xmlData: Document, knownModules: string[]): Diagnostic[] {
+  const diagnostics: Diagnostic[] = []
+  const select = xpath.useNamespaces({ cnxml: NS_CNXML })
+  const documentLinks = select('//cnxml:link[@document]', xmlData) as Node[]
+
+  for (const link of documentLinks) {
+    const linkElement = link as any
+    const linkTargetModule: string = linkElement.getAttribute('document')
+    const [startPosition, endPosition] = calculateElementPositions(linkElement)
+    const preparedDiagnostic: Diagnostic = generateDiagnostic(
+      DiagnosticSeverity.Error,
+      startPosition,
+      endPosition,
+      '',
+      LINK_DIAGNOSTIC_SOURCE
+    )
+
+    if (linkTargetModule == null) {
```

Comment:
Yeah, I thought I had actually added this because of the linter complaining about an explicit `null` check not being there (maybe I had fudged some type thing at the time or something), but removing it now I see it doesn't complain about that anymore. I think it can be a check for empty string instead since the xpath query means the attribute should exist.

Reply:
Oh, snap, I think I remember now -- the linter was griping about needing this check when I was experimenting with `xml2js` before switching to `xmldom`. In any case, it's fixed to be a more appropriate check with test coverage now.